rust-runtime/aws-smithy-http-server/src/routing/mod.rs

```diff
@@ -161,38 +175,46 @@ where
             }
         }
 
-        let status_code = if method_not_allowed {
-            StatusCode::METHOD_NOT_ALLOWED
+        if method_not_allowed {
+            RouterFuture::from_response(
+                Response::builder()
+                    .status(StatusCode::METHOD_NOT_ALLOWED)
+                    .body(crate::body::empty())
+                    .unwrap(),
+            )
         } else {
-            StatusCode::NOT_FOUND
-        };
-        RouterFuture::from_response(
-            Response::builder()
-                .status(status_code)
-                .body(crate::body::empty())
-                .unwrap(),
-        )
+            let error = RuntimeError {
+                protocol: self.protocol,
+                kind: RuntimeErrorKind::UnknownOperation,
+            };
+            RouterFuture::from_response(error.into_response())
+        }
     }
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::protocols::Protocol;
     use crate::{body::boxed, routing::request_spec::*};
     use futures_util::Future;
-    use http::Method;
+    use http::{HeaderMap, Method};
     use std::pin::Pin;
 
     /// Helper function to build a `Request`. Used in other test modules.
-    pub fn req(method: &Method, uri: &str) -> Request<()> {
-        Request::builder().method(method).uri(uri).body(()).unwrap()
+    pub fn req(method: &Method, uri: &str, headers: Option<HeaderMap>) -> Request<()> {
```

Comment:
Use the request builder to add the headers instead of building the request and then modifying its headers.

Reply:
It is functionally identical and inside a test. I prefer it this way to be honest.

Reply:
Also I don't understand how to do what you are suggesting. [Builder](https://docs.rs/http/0.2.6/http/request/struct.Builder.html) doesn't have a method to set a `HeaderMap`, so I would need to do something like:

```rust
       let r = Request::builder().method(method).uri(uri);
       if let Some(headers) = headers.as_ref() {
           for (k, v) in headers.iter() {
               r.header(k, v);
           }
       }
       r.body(()).unwrap()
```

This doesn't compile for different reasons and looks quite ugly.

Am I missing something?